key/rev.py added

diveshuttam opened the pull request:
closes #50

diveshuttam committed:
Rev.py added

closes #50
Modified error message

sourabhtk37 commented on src/cmd/key/rev.sh:
```diff
@@ -27,7 +27,7 @@ Are you sure about this?" || return 1
     sed -i "$revcert" -e "s/^:---/---/"
-    gpg --import "$revcert"
+    call_gpg key/rev.py "$revcert"
     call_fn gpg_send_keys $GPG_KEY
```
There doesn't seem to be a report like "successfully revoked your key".

diveshuttam replied:
It wasn't earlier, I have added `print("Successfully imported revocation certificate")` in python script as I felt it was needed.

diveshuttam replied:
Should I remove it?? If not, any suggestions on the change of message? I am not sure what are the more appropriate terms.

dashohoxha commented on src/gpg/key/rev.py:
```diff
@@ -0,0 +1,17 @@
+            raise Exception("Could not import revocation certificate")
+        else:
+            print("Successfully imported revocation certifiate")
+
```
The user is asking to revoke the existing key. So, this message in my opinion, does not make much sense to the user. The same for the error message.

I would suggest that the python script just return an exit code (0 when successful, anything else if there is an error). The bash script can check this exit code and print a message that is appropriate.

The messages can be like this: "Key revoked successfully." and "Key revocation failed."

dashohoxha replied:
However I think that the message "Key revoked successfully." is redundant. Unix/Linux commands usually don't print any message when everything is Ok.

diveshuttam replied:
Ok returning a status code seems fine to me. gpg does the same, we only issue an error message in  case of a gpgme exception/some other error occurs.

dashohoxha commented:
If the python script is very short, maybe you can execute python code directly from the bash script. For the case of revocation it could be something like this:
```bash
python3 <<EndOfPythonCode
import gpg sys
gpg.Context().op_import(open('$revcert'))
EndOfPythonCode
```

diveshuttam commented:
>If the python script is very short, maybe you can execute python code directly from the bash script

I don't think it would be nice, though it shortens the code, but having python code in one place helps. Suppose I have some common changes in python scripts like adding gpgme exceptions. Or may be some specific changes due to change in GPGME itself. Also my understanding of GPGME is changing as I proceed.  It is nice to have them in one place to correct them. Moreover they don't require much of extra work to make them a file.

sourabhtk37 reviewed (dismissed):
LGTM.

sourabhtk37 commented:
I noticed that if you supply incorrect or revocation cert of another key, then it will not report any error.

diveshuttam commented:
Yup currently revocation certificate path is passed through egpg which in turn generated through 1) fingerprint of the key or 2) is passed by the user. 

In 1) egpg asks for the key ID to gpg only (which will change to gpgme after rewrite). So currently I am basing this pr on the assumption that keyid is correct(because it is being fetched by gpg, which seems reasonable), later the I'll put any checks required in getting fingerprint form gpgme.
In 2) Also the path to revocation certificate is passed by egpg to the python script and exists(with high probability) as it is already being checked in the rev.sh a few lines before calling. (This point is valid to 1 also)

diveshuttam committed:
Add try except checks

Using BaseException as gpgme exceptions are not well documented
will narrow it down when we exaustively test this. Didn't find
any specific exceptions while testing

diveshuttam commented:
@sourabhtk37 I have fixed it using the debug variable, I print the traceback only if `DEBUG='yes'`. This is because long and detailed python tracebacks are not very much useful for normal users. (considering that are using egpg because it is easier). It just prints "Key revocation failed" as of now. Later we may add specific messages like "Internal Error" or "Revocation Certificate not found" etc. based on the exception that is raised.
If one wants to see the traceback and details s/he can use 'egpg set debug yes'.

Currently I am using BaseException as I don't know what exceptions can be raised by gpgme in this scenario. Will narrow this down as we test this. Also could not test any valid exceptions, FileNotFound is valid but unlikely(for revocation certificate) as we are checking it in egpg code only a few lines earlier. (Refer @dashohoxha 's comment on #57)

dashohoxha commented on src/gpg/key/rev.py:
```diff
@@ -1,13 +1,19 @@
+            if result is None:
+                exit(1)
+    except BaseException:
+        if(os.environ['EGPG_DEBUG']=='yes'):
```
Why not use `DEBUG`?

dashohoxha commented on src/egpg.sh:
```diff
@@ -192,6 +192,7 @@ config() {
 
     export GNUPGHOME
     export GPG_TTY=$(tty)
+    export EGPG_DEBUG=$DEBUG
```
Why introduce a new variable?

diveshuttam replied:
I want to test for DEBUG in all python scripts. $DEBUG is not already exported. So did it this way as variable name DEBUG is very common variable used by other programs also, I didn't wanted it to interfere with the already running programs.

UPDATE: I was assuming that these variables get reflected in the main shell. But I was wrong I tried this and saw that calling egpg launches its own bash instance. So $DEBUG will not interfere with the programs from the main shell, I can export DEBUG also.

diveshuttam replied:
I had kind of misinterpreted working of calls in bash, I knew that environments are only propagated downwards but I forgot to consider that each call in Bash launches a new subprocess and hence exporting a variable won't affect the main shell. Have corrected this and now  we are using variable $DEBUG only.

dashohoxha commented on src/cmd/key/rev.sh:
```diff
@@ -27,7 +27,11 @@ Are you sure about this?" || return 1
+    call_gpg key/rev.py "$revcert"
+    
+    local err=$?
+    [[ $err == 0 ]] || fail "Key revokation failed"
```
It should be `revocation`.

diveshuttam committed:
Minor updates

Use DEBUG instead of EGPG_DEBUG
Correct spelling in rev.py

diveshuttam commented:
Are the changes ok??

dashohoxha reviewed (approved):
Sorry. I though this was approved (either by me or by T K Sourabh). Actually when new commits are added approval expires and a new approval is needed.

These changes seem Ok, you can merge them.
